### PR TITLE
Add feature to map output solution to new geometry

### PR DESCRIPTION
**Pull request by ndem0**
· remove useless inheritance from *handler classes
· rewrite filehandler allocator to return the specialized handler
· split write and parse method to get output or cells and points
· add class to handle stl file
· add class to map output to new geometry
· clean code

There are a lot of changes: first of all, i had to change old layout for file handler classes to get specific output or points and cells from a file; in this way, I could implement mapper class. Also, I clean *handler design to make userful the generic FileHandler class (look at allocator for this class): I think is a nice trick to avoid checks everywhere and make code really extendibile (now, it's very easy add support for new file, you just need to create new class and change one line in FileHandler class, and everything should run as before with the new files).

### CODE REVIEW

File: ezyrb/interpolation.py [hunk 1]
```diff
@@ -23,15 +24,15 @@ class Interp(object):
 	:param string output_name: name of the variable (or output) we want to
 		extract from the solution file.
 	:param string namefile_prefix: path and prefix of the solution files.
```
Comment (mtezzele) [inline ezyrb/interpolation.py]: Update the documentation.

File: ezyrb/interpolation.py [hunk 2]
```diff
@@ -185,7 +169,7 @@ def print_info(self):
 			self.cvt_handler.get_optimal_new_mu()
 		)
 
-	def add_snapshot(self, new_mu=None, snapshot_file=None):
+	def add_snapshot(self, new_mu, snapshot_file):
```
Comment (mtezzele) [inline ezyrb/interpolation.py]: Update the documentation.

File: ezyrb/matlabhandler.py [hunk 3]
```diff
@@ -1,67 +1,63 @@
...
 	"""
-	Matlab format file handler class
+	Matlab format file handler class.
+	You are NOT supposed to call directly this class constructor (use
+			Filehandler constructor instead)
 
-	:cvar string infile: name of the input file to be processed.
-	:cvar string extension: extension of the input/output files. It is equal to '.mat'.
-	:cvar string output_name: name of the output of interest inside the mat file. Default value is output.
+	:cvar private _filename: name of file to handler
```
Comment (mtezzele) [inline ezyrb/matlabhandler.py]: With private we are missing the type of the variable. I have not see the result of the compiled doc yet. Perhaps using string after cvar and then describing the member as private in the description would solve the issue.

File: ezyrb/matlabhandler.py [hunk 4]
```diff
@@ -1,67 +1,63 @@
 """
```
Comment (mtezzele) [inline ezyrb/matlabhandler.py]: Update the documentation.

File: ezyrb/matlabhandler.py [hunk 5]
```diff
@@ -1,67 +1,63 @@
...
+	Matlab format file handler class.
+	You are NOT supposed to call directly this class constructor (use
+			Filehandler constructor instead)
 
-	:cvar string infile: name of the input file to be processed.
-	:cvar string extension: extension of the input/output files. It is equal to '.mat'.
-	:cvar string output_name: name of the output of interest inside the mat file. Default value is output.
+	:cvar private _filename: name of file to handler
+	:cvar private _cached_data: private attribute to store the
+		last data processed
```
Comment (mtezzele) [inline ezyrb/matlabhandler.py]: Add the information that the default value is None, please.

File: ezyrb/matlabhandler.py [hunk 6]
```diff
@@ -1,67 +1,63 @@
...
-		loaded_struct = sio.loadmat(self.infile)
-		output_values = np.array(loaded_struct[output_name])
+		if not output_name in self._cached_data:
+			raise RuntimeError("File has no " + output_name + " field.")
 
-		return output_values
+		return np.array(self._cached_data[output_name])
 
-	def write(self, output_values, filename):
+	def set_dataset(self, output_values, output_name):
```
Comment (mtezzele) [inline ezyrb/matlabhandler.py]: Update the documentation.

File: ezyrb/mapper.py [hunk 7]
```diff
@@ -0,0 +1,243 @@
+"""
+Utilities for mapping solutions on a different grid
```
Comment (mtezzele) [inline ezyrb/mapper.py]: Please check the english in the documentation because there are a lot of mistakes and nonsense.

File: ezyrb/mapper.py [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+		self._output_name = output_name
+
+	@property
+	def interpolate_function(self):
+		"""
+		The function to interpolate value of a new point using neighbors value.
+		It has to take as arguments two `numpy.array`: the first indicates
+		neighbors value (one row for each value) and the second indicates
+		distance from interpolated point.
```
Comment (mtezzele) [inline ezyrb/mapper.py]: *the distance from the interpolated point.
Please provide a snippet of code as example.

File: ezyrb/mapper.py [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+		This method build structure for nearest-neighbour lookup; points are
+		dataset.
+
+		:param numpy.ndarray points: a n_points-by-3 where are stored
+			coordinates of points by row
+		"""
+		try:
+			self._neighbour_locator = scipy.spatial.cKDTree(points)
+		except:
+			sys.setrecursion(points.shape[0])
```
Comment (mtezzele) [inline ezyrb/mapper.py]: Just a comment on why this line is necessary. Just in case someone wants to erase it ;)

File: ezyrb/mapper.py [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+			self._neighbour_locator = scipy.spatial.cKDTree(points)
+
+	def find_neighbour(self, coordinates):
+		"""
+		This method looks for the nearest neighbors to a given point; the
+		locator have to build before looking for operation.
```
Comment (mtezzele) [inline ezyrb/mapper.py]: I don't get the meaning of the sentence.

File: ezyrb/mapper.py [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+
+		:param coordinates: coordinates of the point to query :return: a
```
Comment (mtezzele) [inline ezyrb/mapper.py]: The type is missing. The return in a new line please.

File: ezyrb/matlabhandler.py [hunk 12]
```diff
@@ -1,67 +1,63 @@
...
 
-	def parse(self, filename, output_name=None):
+		self._filename = filename
+		self._cached_data = None
+
+	def get_dataset(self, output_name):
 		"""
-		Method to parse the `filename`. It returns a vector (matrix with one column) with all the values of the chosen output.
-		If `output_name` is not given it is set to the default value.
+		Method to parse the `filename`. It returns a vector (matrix with one
```
Comment (mtezzele) [inline ezyrb/matlabhandler.py]: Update the doc

File: ezyrb/pod.py [hunk 13]
```diff
@@ -28,23 +28,22 @@ class Pod(object):
 		solution file for the computation of the errors. If the solution files
 		does not contain any weight (like volume or area of the cells) the
 		weight is set to 1 for all the cells.
-	:param string namefile_prefix: path and prefix of the solution files.
-		The files are supposed to be named with the same prefix, plus an
+	:param string snapshot_file_regex: path and prefix of the solution files.
+		The files are supposed to be named with	the same prefix, plus an
```
Comment (mtezzele) [inline ezyrb/pod.py]: I would add here the sentence erased below: For example, in the directory tests/test_datasets/ you can find the files (matlab_00.vtk, matlab_01.vtk, ..., matlab_05.vtk). Or something like this just to be super clear.

File: ezyrb/stlhandler.py [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+class StlHandler(object):
+	"""
+	Vtk format file handler class.
+	You are NOT supposed to call directly this class constructor (use
+						Filehandler constructor instead)
+
+	:cvar string filename: name of file to handle
+	:cvar _cached_data: private attribute to store the last polydata
+		processed
+
```
Comment (mtezzele) [inline ezyrb/stlhandler.py]: Add a snippet of code as example with the following:
:Example:

	>>> import blabla

File: ezyrb/stlhandler.py [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+
+class StlHandler(object):
+	"""
+	Vtk format file handler class.
+	You are NOT supposed to call directly this class constructor (use
+						Filehandler constructor instead)
+
+	:cvar string filename: name of file to handle
+	:cvar _cached_data: private attribute to store the last polydata
```
Comment (mtezzele) [inline ezyrb/stlhandler.py]: Missing type

File: ezyrb/stlhandler.py [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+
+	def _read_polydata(self):
+		"""
+		This private method reads the given `filename` and return a vtkPolyData
+		object containing all informations about file; to avoid useless IO
+		operation on the same file, it stores polydata of the last file parsed
+		and if user ask for this file, the polydata previously stored is
+		returned.
+
+		:param string filename: name of file to read.
```
Comment (mtezzele) [inline ezyrb/stlhandler.py]: Update the doc

File: ezyrb/stlhandler.py [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+		else:
+			writer.SetInputData(data)
+		writer.Write()
+
+	def get_geometry(self, get_cells=False):
+		"""
+		This method reads the given `filename` and returns points and cells of
+		file.
+
+		:param string filename: name of the file to read.
```
Comment (mtezzele) [inline ezyrb/stlhandler.py]: Update the doc

File: ezyrb/stlhandler.py [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+		else:
+			cells = None
+
+		return points, cells
+
+	def set_geometry(self, points, cells, write_bin=False):
+		"""
+		This method save into `filename` a new data defined by `points`
+		
+		:param string filename: file to write
```
Comment (mtezzele) [inline ezyrb/stlhandler.py]: write_bin is missing

File: ezyrb/vtkhandler.py [hunk 19]
```diff
@@ -3,124 +3,236 @@
...
 
-		# TODO: check if the output is a scalar or vector
-		output_values = ns.vtk_to_numpy(extracted_data).reshape((-1, 1))
+	def get_geometry(self, get_cells=False):
+		"""
+		This method reads the given `filename` and returns points and cells of
+		file.
 
-		return output_values
+		:param string filename: name of the file to read.
```
Comment (mtezzele) [inline ezyrb/vtkhandler.py]: Update the doc

File: tests/test_pod.py [hunk 20]
```diff
@@ -370,9 +300,15 @@ def test_pod_write_structures_01_mat(self):
 	'''def test_pod_write_structures_02(self):
 		output_name = 'Pressure'
 		weights_name = 'Weights'
+<<<<<<< HEAD
```
Comment (mtezzele) [inline tests/test_pod.py]: Pay attention! I think they derive from a conflict. Fix all the "conflicts".

File: ezyrb/mapper.py (hunk 7 above)
Comment (ndem0) [inline ezyrb/mapper.py]: Yeah, you're right, I apologize. My english is horrible, and also I wrote the documentation so much (too much) quickly. I'll recheck everything, at least in mapper class, as soon as possible.

**Comment by ndem0** [comment]
Almost everything is fixed; i don't edit yet the pod file because in the next PR I wanna merge pod and interpolation class in order to avoid duplicate code.

File: ezyrb/stlhandler.py [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+			writer.SetInput(data)
+		else:
+			writer.SetInputData(data)
+		writer.Write()
+
+	def get_geometry(self, get_cells=False):
+		"""
+		This method reads the given `filename` and returns points and cells of
+		file.
+
```
Comment (mtezzele) [inline ezyrb/stlhandler.py]: Missing get_cells doc

File: ezyrb/stlhandler.py [hunk 22]
```diff
@@ -0,0 +1,139 @@
...
+			] for i in np.arange(n_cells)]
+		else:
+			cells = None
+
+		return points, cells
+
+	def set_geometry(self, points, cells, write_bin=False):
+		"""
+		This method save into `filename` a new data defined by `points`
+
```
Comment (mtezzele) [inline ezyrb/stlhandler.py]: Missing cells doc

File: ezyrb/vtkhandler.py [hunk 23]
```diff
@@ -3,124 +3,232 @@
...
+		operation on the same file, it stores polydata of the last file parsed
+		and if user ask for this file, the polydata previously stored is
+		returned.
 
-		if output_name is None:
-			output_name = self.output_name
-		else:
-			self._check_filename_type(output_name)
+		:return: data: polydata containing information about file.
+		:rtpe: vtkPolyData
```
Comment (mtezzele) [inline ezyrb/vtkhandler.py]: rtype*

File: ezyrb/vtkhandler.py [hunk 24]
```diff
@@ -3,124 +3,232 @@
...
 		else:
-			extracted_data = extracted_data_cell
-			self.cell_data = True
+			writer.SetInputData(data)
+		writer.Write()
+
+	def get_geometry(self, get_cells=False):
+		"""
+		This method reads the given `filename` and returns points and cells of
+		file.
```
Comment (mtezzele) [inline ezyrb/vtkhandler.py]: get_cells doc missing

File: ezyrb/vtkhandler.py [hunk 25]
```diff
@@ -3,124 +3,232 @@
...
+
+		self._save_polydata(data)
+
+	def get_dataset(self, output_name, datatype='point'):
+		"""
+		This method reads the given `filename` and returns a numpy array
+		containing `output_name` field for each point (if datatype is 'point')
+		or for each cell (if datatype is 'cell').
+		If `output_name` is not a field in file, raise exception.
+
```
Comment (mtezzele) [inline ezyrb/vtkhandler.py]: Missing params doc

File: ezyrb/gui.py [hunk 26]
```diff
@@ -526,12 +525,8 @@ def _main(self):
 		self.root.config(menu=menubar)
 
 	def start(self):
-		"""
-		This method inizializes and starts the GUI. 
-		"""
-		self._main()
-		self.root.mainloop()
 
+		self.root.mainloop()
```
Comment (mtezzele) [inline ezyrb/gui.py]: you erased self.main() here so you have to add it at the end of the file and in the start_gui.py

File: ezyrb/interpolation.py [hunk 27]
```diff
@@ -23,15 +24,15 @@ class Interp(object):
...
+		The files are supposed to be named with	the same prefix, plus an
 		increasing numeration (from 0) in the same order as the parameter
 		points.
 	:param string file_format: format of the solution files.
 
 	:cvar string output_name: name of the variable (or output) we want to
 		extract from the solution file.
 	:cvar string namefile_prefix: path and prefix of the solution files.
-		The files are supposed to be named with the same prefix, plus an
+		The files are supposed to be named with	the same prefix, plus an
```
Comment (mtezzele) [inline ezyrb/interpolation.py]: Please erase the variables not present anymore.

File: ezyrb/interpolation.py [hunk 28]
```diff
@@ -195,31 +179,17 @@ def add_snapshot(self, new_mu=None, snapshot_file=None):
 		:param snapshot_file string
```
Comment (mtezzele) [inline ezyrb/interpolation.py]: Please more verbose here.

Comment (ndem0) [inline ezyrb/interpolation.py]: Pod, Interpolation and Cvt classes will change in next PR, so I would write documentation/examples/tutorial only for the finished design (and avoid to do it two times).
Do you think is a reasonable idea? Maybe is it better merge everything in this pull request? (imho, merging can create more confusion for code review)

Comment (mtezzele) [inline ezyrb/interpolation.py]: Ok fix the gui.py and I will merge.

**Pull request merged**